aiida/orm/data/structure.py

```diff
@@ -832,39 +832,55 @@ def set_pymatgen_structure(self, struct):
             three directions.
         .. note:: Requires the pymatgen module (version >= 3.3.5, usage
             of earlier versions may cause errors).
-        
+
         :raise ValueError: if there are partial occupancies together with spins.
         """
         def build_kind_name(species_and_occu):
             """
-            Build a kind name from a pymatgen Composition, including an
-            additional ordinal if spin is included, e.g. it returns '<element>1'
-            for an atom with spin<0 and '<element>2' for an element with spin>0,
-            ortherwise (no spin) it returns simply '<element>'
+            Build a kind name from a pymatgen Composition, including an additional ordinal if spin is included,
+            e.g. it returns '<element>1' for an atom with spin < 0 and '<element>2' for an element with spin > 0,
+            otherwise (no spin) it returns simply '<element>'
+
             :param specie: a pymatgen specie
-            :return: a string
+            :return: a string representing the kind name
             """
-            has_spin = any([specie.as_dict().get('properties', {}).get('spin', 0) != 0
-                            for specie in species_and_occu.keys()])
+            species = species_and_occu.keys()
+            occupations = species_and_occu.values()
 
-            if has_spin and (len(species_and_occu.items()) > 1
-                             or any([weight != 1.0 for weight in species_and_occu.values()])):
+            has_spin = any([specie.as_dict().get('properties', {}).get('spin', 0) != 0 for specie in species])
+            has_partial_occupancies = (len(occupations) != 1 or occupations[0] != 1.0)
+
+            if has_partial_occupancies and has_spin:
                 raise ValueError('Cannot set partial occupancies and spins at the same time')
 
-            if not has_spin:
-                return None
+            if has_spin:
+
+                # If there is spin, we can only have a single element, otherwise we would have raised above
+                specie = species[0]
+                spin = specie.as_dict().get('properties', {}).get('spin', 0)
 
-            spin = species_and_occu.keys()[0].as_dict().get('properties', {}).get('spin', 0)
+                if spin < 0:
+                    kind_name = specie.symbol + '1'
+                else:
+                    kind_name = specie.symbol + '2'
 
-            if spin < 0:
-                return specie.symbol + '1'
             else:
-                return specie.symbol + '2'
-        
+                kind_name = ''
```

Comment:
No because line 868 is within the scope of another conditional block. The first block deals with the case when there is spin (and therefore there has to be just one specie), otherwise in the other block, there might be a compound kind and I have to build the string iteratively appending all species

Reply:
sorry, I did not see the else...